File: engine/src/MathCursor.Engine/Lexer.cs
using System;
using System.Collections.Generic;

namespace MathCursor.Engine;

// lexer.js — chars → tokens. GÉNÉRIQUE : consulte le VOCABULAIRE pour savoir si un
// lexème est infixe/préfixe/n-aire. Ne nomme aucun opérateur.
internal static class Lexer
{
    private const int SplitCap = 4; // ≤ 2^4 flux ; au-delà, on découpe tout (un seul flux)

    private static bool IsAlpha(char c) => (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
    private static bool IsDigit(char c) => c >= '0' && c <= '9';
    private static bool IsUpper(char c) => c >= 'A' && c <= 'Z';

    // L'autocorrection typographique FR de Word insère U+00A0 (et U+202F en
    // versions récentes) avant « : ; ! ? » — pour le lexer ce sont des espaces.
    private static bool IsSpace(char c) => c == ' ' || c == '\u00A0' || c == '\u202F';

    // ── jonction de tokens collés (JOIN) — premier match gagne ───────────────
    private sealed class JoinRule
    {
        public string[] Left = Array.Empty<string>();
        public string[] Right = Array.Empty<string>();
        public string[] Roles = Array.Empty<string>();
        public bool Sticky;
        public bool CrossSpace;
    }

    private static readonly JoinRule[] Join =
    {
        // « x2 » → puissance SEULE (ADR 2026-06-10 sup-only : l'indice se
        // force avec _, les suites s'écrivent u_1 — contre-cas documenté).
        new() { Left = new[] { "name" }, Right = new[] { "num" }, Roles = new[] { "sup" }, Sticky = true },
        new() { Left = new[] { "close" }, Right = new[] { "num" }, Roles = new[] { "sup" }, Sticky = true },
        new() { Left = new[] { "unit" }, Right = new[] { "num" }, Roles = new[] { "sup" }, Sticky = true },
        new() { Left = new[] { "num" }, Right = new[] { "unit" }, Roles = new[] { "unitOp" }, Sticky = true, CrossSpace = true },
        new() { Left = new[] { "name", "unit", "post" }, Right = new[] { "open" }, Roles = new[] { "apply" } },
        new() { Left = new[] { "num", "name", "close", "post" }, Right = new[] { "name", "open", "func" }, Roles = new[] { "implicit" } },
    };

    public static List<Token> Lex(string src, EngineCulture culture, Func<int, bool>? choices = null, List<int>? runsOut = null)
    {
        var toks = new List<Token>();
        int i = 0;
        bool sawSpace = false;
        int brk = 0, brc = 0;

        char Ch(int idx) => idx >= 0 && idx < src.Length ? src[idx] : '\0';
        bool Spaced(int a, int b) => IsSpace(Ch(a)) || IsSpace(Ch(b));
        void Push(Token t) { t.SpaceBefore = sawSpace; toks.Add(t); sawSpace = false; }
        Token? Last() => toks.Count > 0 ? toks[toks.Count - 1] : null;

        bool StartsWithAt(string needle, int at)
        {
            if (at + needle.Length > src.Length) return false;
            return string.CompareOrdinal(src, at, needle, 0, needle.Length) == 0;
        }

        // classe+pousse UN mot (grec / mot-infixe / forme déclarée / mot-unité / atome).
        // Les alias de la culture sont résolus ICI (match exact) : les tokens
        // portent des Sym canoniques, l'aval (parser/score/render) ne voit
        // jamais un mot alias. Les replis « atome littéral » gardent le mot
        // original (« dans » isolé doit rendre « dans », pas « in »).
        void Word(string w, bool sp)
        {
            string k = culture.Canon(w);
            Vocabulary.Vocab.TryGetValue(k, out var v);
            VocabEntry? g = v is { Shape: "atom" } ? v
                : Vocabulary.Vocab.TryGetValue(w.ToLowerInvariant(), out var gl) ? gl : null;

            // Word AutoCorrect capitalise en début de phrase (« sum » → « Sum ») :
            // repli insensible à la casse pour les MOTS-CLÉS non-atomes ≥ 2
            // lettres (alias compris : « Somme » → somme → sum). Les atomes
            // gardent leur sémantique de casse (Pi → \Pi, R = ensemble) — leur
            // chemin (g) prime.
            if (v == null && g is not { Shape: "atom" } && w.Length >= 2)
            {
                var k2 = culture.Canon(w.ToLowerInvariant());
                if (k2 != k && Vocabulary.Vocab.TryGetValue(k2, out var v2)
                    && v2.Shape is not null and not "atom") // opérateurs seulement : ni atomes (casse signifiante) ni unités (Sym = mot original)
                { k = k2; v = v2; }
            }

            if (g is { Shape: "atom", Alts: { } alts })
                Push(new Token { Kind = "atom", Syms = alts, Coh = g.Coh });
            else if (g is { Shape: "atom" })
                Push(new Token { Kind = "atom", Sym = IsUpper(w[0]) ? g.Upper : g.Lower });
            else if (v is { Shape: "infix" })
            {
                var prev = Last();
                bool binaryPos = prev != null && (prev.Kind is "atom" or "rparen" or "bracket" or "bar" or "postfix");
                if (binaryPos) Push(new Token { Kind = "infix", Sym = k, Spaced = sp });
                else if (v.Unary != null) Push(new Token { Kind = "prefix", Sym = v.Unary, Spaced = sp });
                else Push(new Token { Kind = "atom", Sym = w });
            }
            else if (v is { WordSpace: true } && !IsSpace(Ch(i)))
                Push(new Token { Kind = "atom", Sym = w });
            else if (v is { Shape: { } shape })
                Push(new Token { Kind = shape, Sym = k, Spaced = sp });
            else if (v is { UnitWord: true })
                Push(new Token { Kind = "atom", Sym = w, UnitWord = true });
            else
                Push(new Token { Kind = "atom", Sym = w });
        }

        // plus long mot-clé Splittable présent dans str à la position at.
        static string BestSplittableAt(string str, int at)
        {
            string b = "";
            foreach (var k in Vocabulary.Splittable)
                if (k.Length > b.Length && at + k.Length <= str.Length && string.CompareOrdinal(str, at, k, 0, k.Length) == 0)
                    b = k;
            return b;
        }

        List<string> Decompose(string s)
        {
            var outp = new List<string>();
            int j = 0;
            while (j < s.Length)
            {
                string best = BestSplittableAt(s, j);
                // capitale de début de phrase (Word AutoCorrect) : « VecAB »,
                // « Sinx »… — en TÊTE de run seulement (c'est là que Word
                // capitalise), retenter en abaissant la 1re lettre et émettre
                // la clé minuscule (la capitale est involontaire).
                if (best.Length == 0 && j == 0 && IsUpper(s[0]) && s.Length >= 2)
                    best = BestSplittableAt(char.ToLowerInvariant(s[0]) + s.Substring(1), 0);
                if (best.Length > 0) { outp.Add(best); j += best.Length; }
                else if (IsUpper(s[j]))
                {
                    // suite de MAJUSCULES = UN morceau (paire de points géo :
                    // « vecAB » → vec + AB, pas vec·A·B où « A » mot-unité
                    // bloquerait la jonction — ADR split-distance-cost-vec).
                    int st = j;
                    while (j < s.Length && IsUpper(s[j])) j++;
                    outp.Add(s.Substring(st, j - st));
                }
                else { outp.Add(s[j].ToString()); j++; }
            }
            return outp;
        }

        while (i < src.Length)
        {
            char c = src[i];
            if (IsSpace(c)) { sawSpace = true; i++; continue; }
            // antislash LaTeX avalé devant un mot : \sum ≡ sum, \forall ≡
            // forall — le mot se résout ensuite par vocab + alias (les noms
            // divergents sont des alias génériques : \infty → inf, etc.).
            if (c == '\\' && IsAlpha(Ch(i + 1))) { i++; continue; }
            if (c == '(') { Push(new Token { Kind = "lparen" }); i++; continue; }
            if (c == ')') { Push(new Token { Kind = "rparen" }); i++; continue; }
            if (c == '[' || c == ']') { Push(new Token { Kind = "bracket", Sym = c.ToString() }); brk++; i++; continue; }
            if (c == '{') { Push(new Token { Kind = "lbrace" }); brc++; i++; continue; }
            if (c == '}') { Push(new Token { Kind = "rbrace" }); brc--; i++; continue; }
            if (c == '|') { bool norm = Ch(i + 1) == '|'; Push(new Token { Kind = "bar", Sym = norm ? "norm" : "abs" }); i += norm ? 2 : 1; continue; }
            if (Vocabulary.Sep.TryGetValue(c, out var rank)) { Push(new Token { Kind = "sep", Sym = c.ToString(), Rank = rank }); i++; continue; }

            // unité COMPOSÉE juste après un nombre (5 m/s)
            var pv = Last();
            if (pv is { Num: true })
            {
                string best = "";
                foreach (var k in Units.Compound.Keys)
                    if (k.Length > best.Length && StartsWithAt(k, i)) best = k;
                if (best.Length > 0) { Push(new Token { Kind = "atom", Sym = Units.Compound[best], UnitWord = true }); i += best.Length; continue; }
            }

            if (IsAlpha(c)) // run de lettres
            {
                int st = i; var sb = new System.Text.StringBuilder();
                while (i < src.Length && IsAlpha(src[i])) sb.Append(src[i++]);
                string s = sb.ToString();
                bool sp = Spaced(st - 1, i);
                // raccourcis dots « v... » / « d... » / « c... » (et formes
                // autocorrigées « v… ») : le mot absorbe les points qui
                // suivent si l'ALIAS « mot... » existe (table aliasGeneric,
                // résolue via Canon — pas un mot, d'où le lookahead). Repli
                // minuscule pour l'autocapitalisation Word (« V... »).
                int dotsLen = StartsWithAt("...", i) ? 3 : Ch(i) == '…' ? 1 : 0;
                if (dotsLen > 0
                    && (Vocabulary.Vocab.TryGetValue(culture.Canon(s + "..."), out var dotsV)
                        || Vocabulary.Vocab.TryGetValue(culture.Canon(s.ToLowerInvariant() + "..."), out dotsV))
                    && dotsV.Shape == "atom")
                {
                    Push(new Token { Kind = "atom", Sym = dotsV.Lower });
                    i += dotsLen;
                    continue;
                }
                // canonicaliser AUSSI ici : « existe » contient « xi » (ξ,
                // Splittable) et ne survit au découpage que si l'alias le
                // rend « connu » dans la culture courante.
                bool known = Vocabulary.Vocab.ContainsKey(culture.Canon(s))
                    || (Vocabulary.Vocab.TryGetValue(s.ToLowerInvariant(), out var lv) && lv.Shape == "atom")
                    || (s.Length >= 2 && Vocabulary.Vocab.ContainsKey(culture.Canon(s.ToLowerInvariant())));
                if (!known)
                {
                    var pieces = Decompose(s);
                    bool anySplit = pieces.Count >= 2 && pieces.Exists(p => Vocabulary.Splittable.Contains(p));
                    if (anySplit)
                    {
                        runsOut?.Add(st);
                        if (choices?.Invoke(st) ?? true)
                        {
                            for (int k = 0; k < pieces.Count; k++) Word(pieces[k], k == 0 && sp);
                            continue;
                        }
                    }
                }
                Word(s, sp);
                continue;
            }

            if (IsDigit(c)) // run de chiffres
            {
                var sb = new System.Text.StringBuilder();
                while (i < src.Length && IsDigit(src[i])) sb.Append(src[i++]);
                if (Array.IndexOf(culture.DecimalsIn, Ch(i)) >= 0
                    && !(Ch(i) == ',' && brc > 0) && IsDigit(Ch(i + 1)))
                {
                    i++; sb.Append(culture.DecimalTex);
                    while (i < src.Length && IsDigit(src[i])) sb.Append(src[i++]);
                }
                Push(new Token { Kind = "atom", Sym = sb.ToString(), Num = true });
                continue;
            }

            // opérateur infixe / postfixe / symbole-atome : PLUS-LONG-MATCH (3, 2 puis 1)
            bool matched = false;
            foreach (int len in new[] { 3, 2, 1 })
            {
                if (i + len > src.Length) continue;
                string op = src.Substring(i, len);
                if (!Vocabulary.Vocab.TryGetValue(op, out var v)
                    || (v.Shape != "infix" && v.Shape != "postfix" && v.Shape != "atom")) continue;
                // symbole-atome (« ... » 3 chars, « … » autocorrigé par Word →
                // \ldots) : poussé comme un mot-atome, Sym = LaTeX du Lit.
                if (v.Shape == "atom") { Push(new Token { Kind = "atom", Sym = v.Lower }); i += len; matched = true; break; }
                if (v.Shape == "postfix") { Push(new Token { Kind = "postfix", Sym = op }); i += len; matched = true; break; }

                var prev = Last();
                bool binaryPos = prev != null && (prev.Kind is "atom" or "rparen" or "bracket" or "bar" or "postfix");
                char after = Ch(i + len);
                bool detachedR = IsSpace(after) || ")],;".IndexOf(after) >= 0;
                if (len == 1 && v.PostSign != null && binaryPos && !sawSpace && detachedR)
                {
                    Push(new Token { Kind = "infix", Sym = Vocabulary.Role["sup"], Sticky = true });
                    Push(new Token { Kind = "atom", Sym = v.PostSign });
                }
                else if (len == 1 && v.Unary != null && !binaryPos)
                    Push(new Token { Kind = "prefix", Sym = v.Unary, Spaced = Spaced(i - 1, i + 1) });
                else if (v.Alts != null)
                    Push(new Token { Kind = "infix", Syms = v.Alts, Spaced = Spaced(i - 1, i + len) });
                else
                    Push(new Token { Kind = "infix", Sym = op, Spaced = Spaced(i - 1, i + len) });
                i += len; matched = true; break;
            }
            if (matched) continue;
            throw new Exception("caractère inattendu: " + c);
        }

        // saisie en cours : fermer VIRTUELLEMENT les ( non fermées
        int depth = 0;
        foreach (var t in toks) depth += t.Kind == "lparen" ? 1 : t.Kind == "rparen" ? -1 : 0;
        while (depth-- > 0) toks.Add(new Token { Kind = "rparen", SpaceBefore = false, Virtual = true });
        if (brk % 2 == 1) toks.Add(new Token { Kind = "bracket", Sym = "]", SpaceBefore = false, Virtual = true });

        return Juxtapose(toks);
    }

    private static string? Cat(Token? t)
    {
        if (t == null) return null;
        if (t.Kind == "atom")
        {
            if (t.UnitWord) return "unit";
            bool allDigits = t.Sym != null && t.Sym.Length > 0;
            if (allDigits) foreach (var ch in t.Sym!) if (!IsDigit(ch)) { allDigits = false; break; }
            return t.Num || allDigits ? "num" : "name";
        }
        return t.Kind switch
        {
            "lparen" => "open",
            "rparen" => "close",
            "postfix" => "post",
            "prefix" or "nary" => "func",
            _ => "op",
        };
    }

    private static List<Token> Juxtapose(List<Token> toks)
    {
        var outp = new List<Token>();
        for (int i = 0; i < toks.Count; i++)
        {
            var L = toks[i];
            var R = i + 1 < toks.Count ? toks[i + 1] : null;
            outp.Add(L);
            if (R == null) continue;

            var lc = Cat(L); var rc = Cat(R);
            JoinRule? rule = null;
            foreach (var r in Join)
                if (Array.IndexOf(r.Left, lc) >= 0 && Array.IndexOf(r.Right, rc) >= 0) { rule = r; break; }
            if (rule == null || (R.SpaceBefore && !rule.CrossSpace)) continue;

            var syms = new List<string>();
            foreach (var role in rule.Roles) if (Vocabulary.Role.TryGetValue(role, out var sym)) syms.Add(sym);
            if (syms.Count == 0) continue;

            var t = new Token { Kind = "infix", Spaced = rule.CrossSpace && R.SpaceBefore };
            if (syms.Count > 1) t.Syms = syms; else t.Sym = syms[0];
            if (Array.IndexOf(rule.Roles, "implicit") >= 0 || Array.IndexOf(rule.Roles, "apply") >= 0) t.Implicit = true;
            if (rule.Sticky) t.Sticky = true;
            outp.Add(t);
        }
        return outp;
    }

    private static int PopCount(int m) { int c = 0; while (m != 0) { c += m & 1; m >>= 1; } return c; }

    // NIVEAU 2 — lexAll : tous les flux (run entier OU découpé), tagués du nb de découpes.
    public static List<(List<Token> Toks, int Splits)> LexAll(string src, EngineCulture culture)
    {
        var runs = new List<int>();
        var whole = Lex(src, culture, _ => false, runs);
        int n = runs.Count;
        if (n == 0) return new() { (whole, 0) };
        if (n > SplitCap) return new() { (Lex(src, culture, _ => true), n), (whole, 0) };
        var outp = new List<(List<Token>, int)>();
        for (int m = 0; m < (1 << n); m++)
        {
            int mm = m;
            outp.Add((Lex(src, culture, st => (mm & (1 << runs.IndexOf(st))) != 0), PopCount(mm)));
        }
        return outp;
    }
}

File: engine-python/mc_engine/lexer.py
"""chars -> tokens — port de Lexer.cs.

Générique : consulte le VOCABULAIRE (data) pour savoir si un lexème est
infixe/préfixe/n-aire. Ne nomme aucun opérateur.
"""
from .node import Token
from .vocab import VOCAB, SPLITTABLE, ROLE, SEP, UNITS_COMPOUND

SPLIT_CAP = 4  # <= 2^4 flux ; au-delà, on découpe tout (un seul flux)


def _is_alpha(c): return ('a' <= c <= 'z') or ('A' <= c <= 'Z')
def _is_digit(c): return '0' <= c <= '9'
def _is_upper(c): return 'A' <= c <= 'Z'
# autocorrection FR de Word : U+00A0 / U+202F = espaces pour le lexer.
def _is_space(c): return c in (' ', ' ', ' ')


# jonction de tokens collés (JOIN) — premier match gagne.
JOIN = [
    {"left": ["name"], "right": ["num"], "roles": ["sup"], "sticky": True, "cross": False},
    {"left": ["close"], "right": ["num"], "roles": ["sup"], "sticky": True, "cross": False},
    {"left": ["unit"], "right": ["num"], "roles": ["sup"], "sticky": True, "cross": False},
    {"left": ["num"], "right": ["unit"], "roles": ["unitOp"], "sticky": True, "cross": True},
    {"left": ["name", "unit", "post"], "right": ["open"], "roles": ["apply"], "sticky": False, "cross": False},
    {"left": ["num", "name", "close", "post"], "right": ["name", "open", "func"], "roles": ["implicit"], "sticky": False, "cross": False},
]


def lex(src, culture, choices=None, runs_out=None):
    toks = []
    i = 0
    saw_space = False
    brk = 0
    brc = 0

    def ch(idx):
        return src[idx] if 0 <= idx < len(src) else '\0'

    def spaced(a, b):
        return _is_space(ch(a)) or _is_space(ch(b))

    def push(t):
        nonlocal saw_space
        t.space_before = saw_space
        toks.append(t)
        saw_space = False

    def last():
        return toks[-1] if toks else None

    def word(w, sp):
        k = culture.canon(w)
        v = VOCAB.get(k)
        g = v if (v is not None and v.shape == "atom") else VOCAB.get(w.lower())
        # repli casse (Word AutoCorrect « Sum » -> sum) : mots-clés non-atomes >= 2.
        if v is None and not (g is not None and g.shape == "atom") and len(w) >= 2:
            k2 = culture.canon(w.lower())
            v2 = VOCAB.get(k2)
            if k2 != k and v2 is not None and v2.shape is not None and v2.shape != "atom":
                k = k2
                v = v2

        if g is not None and g.shape == "atom" and g.alts is not None:
            push(Token(kind="atom", syms=list(g.alts), coh=g.coh))
        elif g is not None and g.shape == "atom":
            push(Token(kind="atom", sym=(g.upper if _is_upper(w[0]) else g.lower)))
        elif v is not None and v.shape == "infix":
            prev = last()
            binary = prev is not None and prev.kind in ("atom", "rparen", "bracket", "bar", "postfix")
            if binary:
                push(Token(kind="infix", sym=k, spaced=sp))
            elif v.unary is not None:
                push(Token(kind="prefix", sym=v.unary, spaced=sp))
            else:
                push(Token(kind="atom", sym=w))
        elif v is not None and v.word_space and not _is_space(ch(i)):
            push(Token(kind="atom", sym=w))
        elif v is not None and v.shape is not None:
            push(Token(kind=v.shape, sym=k, spaced=sp))
        elif v is not None and v.unit_word:
            push(Token(kind="atom", sym=w, unit_word=True))
        else:
            push(Token(kind="atom", sym=w))

    def best_splittable_at(s, at):
        b = ""
        for key in SPLITTABLE:
            if len(key) > len(b) and at + len(key) <= len(s) and s.startswith(key, at):
                b = key
        return b

    def decompose(s):
        outp = []
        j = 0
        while j < len(s):
            best = best_splittable_at(s, j)
            if best == "" and j == 0 and _is_upper(s[0]) and len(s) >= 2:
                best = best_splittable_at(s[0].lower() + s[1:], 0)
            if best:
                outp.append(best); j += len(best)
            elif _is_upper(s[j]):
                st = j
                while j < len(s) and _is_upper(s[j]):
                    j += 1
                outp.append(s[st:j])
            else:
                outp.append(s[j]); j += 1
        return outp

    while i < len(src):
        c = src[i]
        if _is_space(c):
            saw_space = True; i += 1; continue
        if c == '\\' and _is_alpha(ch(i + 1)):
            i += 1; continue
        if c == '(':
            push(Token(kind="lparen")); i += 1; continue
        if c == ')':
            push(Token(kind="rparen")); i += 1; continue
        if c == '[' or c == ']':
            push(Token(kind="bracket", sym=c)); brk += 1; i += 1; continue
        if c == '{':
            push(Token(kind="lbrace")); brc += 1; i += 1; continue
        if c == '}':
            push(Token(kind="rbrace")); brc -= 1; i += 1; continue
        if c == '|':
            norm = ch(i + 1) == '|'
            push(Token(kind="bar", sym="norm" if norm else "abs")); i += 2 if norm else 1; continue
        if c in SEP:
            push(Token(kind="sep", sym=c, rank=SEP[c])); i += 1; continue

        # raccourcis grecs « @ » + lettre(s)
        if c == '@' and _is_alpha(ch(i + 1)):
            at0 = i; i += 1
            gr = ""
            while i < len(src) and _is_alpha(src[i]):
                gr += src[i]; i += 1
            gsp = spaced(at0 - 1, i)
            if len(gr) == 1:
                up = _is_upper(gr[0])
                low = "@" + gr[0].lower()
                key = culture.canon(low)
                if key != low:
                    ve = VOCAB[key]
                    if ve.alts is not None:
                        alts = ve.alts_upper if up else ve.alts
                        if alts is not None:
                            push(Token(kind="atom", syms=list(alts))); continue
                    elif (not up) or (ve.upper != ve.lower):
                        push(Token(kind="atom", sym=(ve.upper if up else ve.lower))); continue
            word(gr, gsp)
            continue

        # unité COMPOSÉE juste après un nombre (5 m/s)
        pv = last()
        if pv is not None and pv.num:
            best = ""
            for key in UNITS_COMPOUND:
                if len(key) > len(best) and src.startswith(key, i):
                    best = key
            if best:
                push(Token(kind="atom", sym=UNITS_COMPOUND[best], unit_word=True)); i += len(best); continue

        if _is_alpha(c):  # run de lettres
            st = i; s = ""
            while i < len(src) and _is_alpha(src[i]):
                s += src[i]; i += 1
            sp = spaced(st - 1, i)
            dots_len = 3 if src.startswith("...", i) else (1 if ch(i) == '…' else 0)
            if dots_len > 0:
                dv = VOCAB.get(culture.canon(s + "...")) or VOCAB.get(culture.canon(s.lower() + "..."))
                if dv is not None and dv.shape == "atom":
                    push(Token(kind="atom", sym=dv.lower)); i += dots_len; continue
            known = (culture.canon(s) in VOCAB) \
                or (VOCAB.get(s.lower()) is not None and VOCAB[s.lower()].shape == "atom") \
                or (len(s) >= 2 and culture.canon(s.lower()) in VOCAB)
            if not known:
                pieces = decompose(s)
                any_split = len(pieces) >= 2 and any(p in SPLITTABLE for p in pieces)
                if any_split:
                    if runs_out is not None:
                        runs_out.append(st)
                    if (choices(st) if choices is not None else True):
                        for kk in range(len(pieces)):
                            word(pieces[kk], kk == 0 and sp)
                        continue
            word(s, sp)
            continue

        if _is_digit(c):  # run de chiffres
            s = ""
            while i < len(src) and _is_digit(src[i]):
                s += src[i]; i += 1
            if ch(i) in culture.decimals_in and not (ch(i) == ',' and brc > 0) and _is_digit(ch(i + 1)):
                i += 1; s += culture.decimal_tex
                while i < len(src) and _is_digit(src[i]):
                    s += src[i]; i += 1
            push(Token(kind="atom", sym=s, num=True)); continue

        # opérateur infixe / postfixe / symbole-atome : PLUS-LONG-MATCH (3, 2, 1)
        matched = False
        for length in (3, 2, 1):
            if i + length > len(src):
                continue
            op = src[i:i + length]
            v = VOCAB.get(op)
            if v is None or v.shape not in ("infix", "postfix", "atom"):
                continue
            if v.shape == "atom":
                push(Token(kind="atom", sym=v.lower)); i += length; matched = True; break
            if v.shape == "postfix":
                push(Token(kind="postfix", sym=op)); i += length; matched = True; break
            prev = last()
            binary = prev is not None and prev.kind in ("atom", "rparen", "bracket", "bar", "postfix")
            after = ch(i + length)
            detached_r = _is_space(after) or after in ")],;"
            if length == 1 and v.post_sign is not None and binary and not saw_space and detached_r:
                push(Token(kind="infix", sym=ROLE["sup"], sticky=True))
                push(Token(kind="atom", sym=v.post_sign))
            elif length == 1 and v.unary is not None and not binary:
                push(Token(kind="prefix", sym=v.unary, spaced=spaced(i - 1, i + 1)))
            elif v.alts is not None:
                push(Token(kind="infix", syms=list(v.alts), spaced=spaced(i - 1, i + length)))
            else:
                push(Token(kind="infix", sym=op, spaced=spaced(i - 1, i + length)))
            i += length; matched = True; break
        if matched:
            continue
        raise Exception("caractère inattendu: " + c)

    # saisie en cours : fermer VIRTUELLEMENT les ( non fermées
    depth = 0
    for t in toks:
        depth += 1 if t.kind == "lparen" else (-1 if t.kind == "rparen" else 0)
    while depth > 0:
        toks.append(Token(kind="rparen", space_before=False, virtual=True)); depth -= 1
    if brk % 2 == 1:
        toks.append(Token(kind="bracket", sym="]", space_before=False, virtual=True))

    return _juxtapose(toks)


def _cat(t):
    if t is None:
        return None
    if t.kind == "atom":
        if t.unit_word:
            return "unit"
        all_digits = t.sym is not None and len(t.sym) > 0 and all(_is_digit(x) for x in t.sym)
        return "num" if (t.num or all_digits) else "name"
    return {"lparen": "open", "rparen": "close", "postfix": "post",
            "prefix": "func", "nary": "func"}.get(t.kind, "op")


def _juxtapose(toks):
    outp = []
    for idx in range(len(toks)):
        L = toks[idx]
        R = toks[idx + 1] if idx + 1 < len(toks) else None
        outp.append(L)
        if R is None:
            continue
        lc = _cat(L); rc = _cat(R)
        rule = None
        for r in JOIN:
            if lc in r["left"] and rc in r["right"]:
                rule = r; break
        if rule is None or (R.space_before and not rule["cross"]):
            continue
        syms = [ROLE[role] for role in rule["roles"] if role in ROLE]
        if not syms:
            continue
        t = Token(kind="infix", spaced=(rule["cross"] and R.space_before))
        if len(syms) > 1:
            t.syms = syms
        else:
            t.sym = syms[0]
        if "implicit" in rule["roles"] or "apply" in rule["roles"]:
            t.implicit = True
        if rule["sticky"]:
            t.sticky = True
        outp.append(t)
    return outp


def _popcount(m):
    return bin(m).count("1")


def lex_all(src, culture):
    runs = []
    whole = lex(src, culture, lambda st: False, runs)
    n = len(runs)
    if n == 0:
        return [(whole, 0)]
    if n > SPLIT_CAP:
        return [(lex(src, culture, lambda st: True), n), (whole, 0)]
    outp = []
    for m in range(1 << n):
        outp.append((lex(src, culture, lambda st, mm=m: (mm & (1 << runs.index(st))) != 0), _popcount(m)))
    return outp
